### PR TITLE
Handle _selected_range sent to NSTextInputClient.setMarkedText().

### PR DESCRIPTION
**Pull request by ShikiSuen**
- [ ] Tested on all platforms changed // Only macOS.
- [ ] Added an entry to the `changelog` module if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

This only solves the `_selected_range` part of this issue: https://github.com/rust-windowing/winit/issues/3617

### CODE REVIEW

**Comment by ShikiSuen** [comment]
Some CI typo checks might need improvements...

File: src/platform_impl/macos/view.rs
Comment (madsmtm) [inline src/platform_impl/macos/view.rs:281]: Nit: Feels overly verbose, maybe just:
```suggestion
            // TODO: Use `_replacement_range` correctly (known to be used by Japanese IMEs).
```

File: src/platform_impl/macos/view.rs
Comment (madsmtm) [inline src/platform_impl/macos/view.rs:340]: I feel like there are ways of making this code easier to read.

1. Maybe don't eagerly convert the `preedit_string` to a Rust `String` on line `288`/`296` in original code, but instead do something like the following, to get a `Id<NSString>` instead:
    ```
    // Line 288
    string.string(),
    // Line 296
    string.retain(),
    ```

    That way you at least won't have to convert from `&str` to `NSString` inside of `get_actual_range_bounds`, which will be more efficient.

2. Move the `!preedit_string.is_empty()` down into the `else` branch here, and get rid of the intermediary `lowerbound`/`upperbound` values.

File: src/platform_impl/macos/view.rs
Comment (madsmtm) [inline src/platform_impl/macos/view.rs:1151]: From just reading the code, I cannot tell what this code does, maybe describe it more in words, that it converts from the UTF-16 that AppKit uses in its NSRange to the UTF-8 that Rust uses.

And why is it marked as `unsafe`? Maybe it'll be more clear if inlined instead of split into a separate function?

**Review by madsmtm — CHANGES_REQUESTED** [review]
Thanks, I've been thinking about using these values as well in the past, but ended up forgetting about it exactly because the UTF-16 to UTF-8 conversion was a bit tricky, and I was lazy.

File: src/platform_impl/macos/view.rs
Comment (pan93412) [inline src/platform_impl/macos/view.rs:333]: I think this statement can be simplified with `if-let` syntax, as the `None` branch is empty.

```suggestion
                if let Some(fetched_value) = new_bounds {
	            lowerbound = fetched_value.0;
	            upperbound = fetched_value.1;
                }
```

File: src/platform_impl/macos/view.rs
Comment (pan93412) [inline src/platform_impl/macos/view.rs:1170]: I think we do not need to make the meaningless allocation if we do not need.

CleanShot 2024-04-03 at 13 03 58@2x


```suggestion
    let lowerbound_u8: usize = sub_string_a.len();
    let upperbound_u8: usize = sub_string_b.len();
```

**Review by pan93412 — CHANGES_REQUESTED** [review]

File: src/platform_impl/macos/view.rs
Comment (ShikiSuen) [reply src/platform_impl/macos/view.rs:1145]: @madsmtm Thanks. Pan93412 also has the same opinion. I just fixed this.

File: src/platform_impl/macos/view.rs
Comment (ShikiSuen) [reply src/platform_impl/macos/view.rs]: I think it is good. I am just afraid that I am just an amateur FNG dealing with Rust, and I don't want to screw it up. I have more experience with Swift but less with ObjC (and all pointer-related things).

This PR is editable by maintainers. Could you please help me apply your suggesions with your commits?

**Comment by ShikiSuen** [comment]
Thanks to @pan93412 and @madsmtm 's suggestions. Most of these suggestions are applied on web, hence the spammed commit history. I apologize if that brings any inconvenience.

**Comment by ShikiSuen** [comment]
Memo: I'm revising the sanity check of the get_actual_range_bounds().
Maybe Pan misunderstood my intention here. I need to add some more inline commments.

**Comment by ShikiSuen** [comment]
The revision is done. I guess this PR is ready for merge.
Or we can wait for @madsmtm's refactor commit if he is glad to provide one.

P.S.: I just don't know how to handle the `Id<T>` in Rust...

**Comment by pan93412** [comment]
It may be better to squash all the commits in this PR to one.

**Comment by ShikiSuen** [comment]
@pan93412 Thanks. Squashed.

This PR needs to be force-pulled due to the force-pushed suqashed commit. cc @madsmtm

File: src/platform_impl/macos/view.rs
Comment (pan93412) [inline src/platform_impl/macos/view.rs:1146]: As the logic of this code is obvious, I think you don't need to describe it in your documentation again.

Comment (ShikiSuen) [reply src/platform_impl/macos/view.rs:1146]: That's requested by @madsmtm .

File: src/platform_impl/macos/view.rs
Comment (kchibisov) [inline src/platform_impl/macos/view.rs:394]: Just remove all this comment all together, there's no API to even utilize the `replacement_range`.

If you don't really want to do so, you can comment like `/// TODO: Requires surrounding text reporting on the Window.`.

File: src/platform_impl/macos/view.rs
Comment (kchibisov) [inline src/platform_impl/macos/view.rs:322]: You can merge all of that into a single `if let`.

File: src/platform_impl/macos/view.rs
Comment (kchibisov) [inline src/platform_impl/macos/view.rs:279]: Same comment as below for your _bigger_ todo, I'd suggest to just keep one of them present, likely right above the `_replacement_range`

**Review by kchibisov — CHANGES_REQUESTED** [review]

File: src/platform_impl/macos/view.rs
Comment (ShikiSuen) [reply src/platform_impl/macos/view.rs:394]: > there's no API to even utilize the replacement_range

Prejudices are not professional at all. This parameter is supported in Qt:
https://github.com/qt/qtbase/blob/3941fe697b9fea633f85321031ece765c532d46d/src/plugins/platforms/cocoa/qnsview_complextext.mm#L41-L93

File: src/platform_impl/macos/view.rs
Comment (pan93412) [inline src/platform_impl/macos/view.rs:1142]:
```suggestion
    // The given NSRange length may be 0 to indicate a cursor in line of a marked range.
```

typo?

File: src/platform_impl/macos/view.rs
Comment (pan93412) [inline src/platform_impl/macos/view.rs:1149]:
```suggestion
    //    TODO: rewrite this code if any of them wants to handle it like that
```

File: src/platform_impl/macos/view.rs
Comment (pan93412) [inline src/platform_impl/macos/view.rs:1155]:
```suggestion
    // Sanity Check. Return None if it fails.
```

File: src/platform_impl/macos/view.rs
Comment (pan93412) [inline src/platform_impl/macos/view.rs:1166]:
```suggestion
    // 4. Use the two new sliced NSStrings' rust len() as the returned calculated UTF-8 bounds.
```

File: src/platform_impl/macos/view.rs
Comment (kchibisov) [reply src/platform_impl/macos/view.rs]: No API in **winit**.

Comment (ShikiSuen) [reply src/platform_impl/macos/view.rs]: > No API in **winit**.

Then supply the API later.

File: src/platform_impl/macos/view.rs
Comment (pan93412) [inline src/platform_impl/macos/view.rs:324]: I would prefer this solution:

```suggestion
            let (lowerbound, upperbound) = if let Some(fetched_value) = get_actual_range_bounds(&preedit_string, selected_range).unwrap_or_else(|| (preedit_string.len(), lowerbound));
```

I did not test if this passes the compilation.

**Review by pan93412 — CHANGES_REQUESTED** [review]

File: src/platform_impl/macos/view.rs
Comment (ShikiSuen) [reply src/platform_impl/macos/view.rs:1142]: Google it. I learned this expression from Dr. Ken Lunde.

File: src/platform_impl/macos/view.rs
Comment (kchibisov) [reply src/platform_impl/macos/view.rs]: Yes, but there's no point in `TODO` in code when you can have all of that in the proper issues. No one really looks in the code TODOs. The only exception I'd say use that if it's for yourself, which is not the case mostly with most TODOs.

File: src/platform_impl/macos/view.rs
Comment (ShikiSuen) [reply src/platform_impl/macos/view.rs]: Thanks for your suggestion. `.unwrap_or(fallbackValue)` is enough in this case.

**Review by kchibisov — APPROVED** [review]
r @madsmtm

**Review by pan93412 — APPROVED** [review]

**Comment by kchibisov** [comment]
Besides the @madsmtm review. You should pass the `CI` (use nightly formatter).

**Comment by ShikiSuen** [comment]
@kchibisov 
image

**Comment by pan93412** [comment]
> @kchibisov 
> 
> image
> 
> 

You may install it with `rustup toolchain install nightly`: https://rust-lang.github.io/rustup/concepts/channels.html

**Comment by ShikiSuen** [comment]
@pan93412 Sry but seems too late. Something seems to be permanently crapped on my computer due to the installation of the nightly.

I dunno why depending on nightly CI utilities for stability-critical scenarios.

```
~/Repos/winit > cargo install rustfmt-nightly                                           05/04/2024 02:00:22 PM
    Updating crates.io index
  Downloaded rustfmt-nightly v1.4.21
  Downloaded 1 crate (519.2 KB) in 1.53s
error: binary `cargo-fmt` already exists in destination
binary `rustfmt` already exists in destination
Add --force to overwrite
~/Repos/winit > cargo +nightly fmt                                                      05/04/2024 02:00:28 PM
error: toolchain 'nightly-x86_64-apple-darwin' is not installed
~/Repos/winit > cargo install rustfmt-nightly --force                                   05/04/2024 02:00:36 PM
    Updating crates.io index
  Installing rustfmt-nightly v1.4.21
    Updating crates.io index
  Downloaded autocfg v1.3.0
  Downloaded byte-tools v0.3.1
  Downloaded block-padding v0.1.5
  Downloaded vec_map v0.8.2
  Downloaded fake-simd v0.1.2
  Downloaded opaque-debug v0.2.3
  Downloaded block-buffer v0.7.3
  Downloaded ansi_term v0.11.0
  Downloaded arrayvec v0.5.2
  Downloaded dirs-sys v0.3.7
  Downloaded dirs v2.0.2
  Downloaded digest v0.8.1
  Downloaded fastrand v2.1.0
  Downloaded humantime v1.3.0
  Downloaded lock_api v0.3.4
  Downloaded ena v0.14.2
  Downloaded generic-array v0.12.4
  Downloaded derive-new v0.5.9
  Downloaded rustc-ap-rustc_lexer v677.0.0
  Downloaded rustc-ap-rustc_feature v677.0.0
  Downloaded psm v0.1.21
  Downloaded semver v0.9.0
  Downloaded scoped-tls v1.0.1
  Downloaded textwrap v0.11.0
  Downloaded thiserror-impl v1.0.59
  Downloaded stacker v0.1.15
  Downloaded synstructure v0.12.6
  Downloaded unicode-xid v0.2.4
  Downloaded unicode-width v0.1.12
  Downloaded semver-parser v0.7.0
  Downloaded anyhow v1.0.82
  Downloaded env_logger v0.6.2
  Downloaded quick-error v1.2.3
  Downloaded maybe-uninit v2.0.0
  Downloaded parking_lot_core v0.6.3
  Downloaded quote v1.0.36
  Downloaded memmap v0.7.0
  Downloaded thiserror v1.0.59
  Downloaded tracing-attributes v0.1.27
  Downloaded proc-macro2 v1.0.81
  Downloaded ignore v0.4.22
  Downloaded typenum v1.17.0
  Downloaded structopt v0.3.26
  Downloaded toml v0.5.11
  Downloaded serde v1.0.200
  Downloaded itertools v0.8.2
  Downloaded unicode_categories v0.1.1
  Downloaded cc v1.0.96
  Downloaded rustc-ap-rustc_parse v677.0.0
  Downloaded unicode-normalization v0.1.23
  Downloaded serde_derive v1.0.200
  Downloaded serde_json v1.0.116
  Downloaded clap v2.34.0
  Downloaded rustc-rayon v0.3.2
  Downloaded syn v2.0.60
  Downloaded rustix v0.38.34
  Downloaded term v0.6.1
  Downloaded rustc-ap-rustc_target v677.0.0
  Downloaded rustc-ap-rustc_data_structures v677.0.0
  Downloaded rustc-ap-rustc_ast v677.0.0
  Downloaded rustc-ap-rustc_expand v677.0.0
  Downloaded rustfmt-config_proc_macro v0.2.0
  Downloaded rustc-workspace-hack v1.0.0
  Downloaded rustc-ap-rustc_span v677.0.0
  Downloaded rustc-ap-rustc_serialize v677.0.0
  Downloaded rustc-ap-rustc_macros v677.0.0
  Downloaded rustc-ap-rustc_index v677.0.0
  Downloaded rustc-ap-rustc_graphviz v677.0.0
  Downloaded rustc-ap-rustc_fs_util v677.0.0
  Downloaded rustc-ap-rustc_errors v677.0.0
  Downloaded rustc-ap-rustc_attr v677.0.0
  Downloaded rustc-ap-rustc_ast_pretty v677.0.0
  Downloaded rustc-ap-rustc_ast_passes v677.0.0
  Downloaded rustc-ap-rustc_arena v677.0.0
  Downloaded parking_lot_core v0.7.3
  Downloaded measureme v0.7.1
  Downloaded md-5 v0.8.0
  Downloaded libc v0.2.154
  Downloaded cargo_metadata v0.8.2
  Downloaded bstr v1.9.1
  Downloaded annotate-snippets v0.6.1
  Downloaded rustc-rayon-core v0.3.2
  Downloaded rustc-ap-rustc_session v677.0.0
  Downloaded smallvec v0.6.14
  Downloaded diff v0.1.13
  Downloaded crossbeam-utils v0.7.2
  Downloaded crossbeam-epoch v0.9.18
  Downloaded annotate-snippets v0.8.0
  Downloaded parking_lot v0.10.2
  Downloaded parking_lot v0.9.0
  Downloaded jobserver v0.1.31
  Downloaded globset v0.4.14
  Downloaded either v1.11.0
  Downloaded termize v0.1.1
  Downloaded termcolor v1.4.1
  Downloaded rustc_version v0.2.3
  Downloaded structopt-derive v0.4.18
  Downloaded strsim v0.8.0
  Downloaded sha-1 v0.8.2
  Downloaded crossbeam-deque v0.8.5
  Downloaded bytecount v0.6.8
  Downloaded 101 crates (5.1 MB) in 1.54s
   Compiling proc-macro2 v1.0.81
   Compiling unicode-ident v1.0.12
   Compiling libc v0.2.154
   Compiling serde v1.0.200
   Compiling syn v1.0.109
   Compiling semver-parser v0.7.0
   Compiling autocfg v1.3.0
   Compiling cfg-if v0.1.10
   Compiling crossbeam-utils v0.8.19
   Compiling cc v1.0.96
   Compiling maybe-uninit v2.0.0
   Compiling typenum v1.17.0
   Compiling scopeguard v1.2.0
   Compiling lazy_static v1.4.0
   Compiling byteorder v1.5.0
   Compiling smallvec v1.13.2
   Compiling log v0.4.21
   Compiling smallvec v0.6.14
error[E0554]: `#![feature]` may not be used on the stable release channel
  --> /Users/shikisuen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/smallvec-1.13.2/src/lib.rs:96:37
   |
96 | #![cfg_attr(feature = "may_dangle", feature(dropck_eyepatch))]
   |                                     ^^^^^^^^^^^^^^^^^^^^^^^^

   Compiling lock_api v0.3.4
   Compiling hashbrown v0.12.3
error[E0557]: feature has been removed
  --> /Users/shikisuen/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lock_api-0.3.4/src/lib.rs:91:42
   |
91 | #![cfg_attr(feature = "nightly", feature(const_fn))]
   |                                          ^^^^^^^^ feature has been removed
   |
   = note: split into finer-grained feature gates

For more information about this error, try `rustc --explain E0554`.
error: could not compile `smallvec` (lib) due to previous error
warning: build failed, waiting for other jobs to finish...
For more information about this error, try `rustc --explain E0557`.
error: could not compile `lock_api` (lib) due to previous error
error: failed to compile `rustfmt-nightly v1.4.21`, intermediate artifacts can be found at `/var/folders/10/6wq7rrd57qx334gglsl_cfs00000gq/T/cargo-installghlNPo`.
To reuse those artifacts with a future compilation, set the environment variable `CARGO_TARGET_DIR` to that path.
~/Repos/winit >                                                                         05/04/2024 02:00:45 PM

```

**Comment by ShikiSuen** [comment]
@pan93412 I managed to install the nightly toolchain but still got this after running `cargo fmt`.

I can't understand why we have to waste time on unstable toolchains.

image

**Comment by ShikiSuen** [comment]
I'm out. Will back whenever political correctness becomes less important than fixing crucial bugs.

**Comment by ShikiSuen** [comment]
@kchibisov The nightly rustfmt hinders the TODO to be in a standalone line. Is there a way to stop the nightly rustfmt from doing this stupid?

image

**Comment by pan93412** [comment]
> @pan93412 I managed to install the nightly toolchain but still got this after running `cargo fmt`.
> 
> I can't understand why we have to waste time on unstable toolchains.

To pick up the nightly toolchain, run `cargo +nightly fmt`. You can switch your default toolchain to `nightly` by running `rustup default nightly`.

Some Rust users prefer the `nightly` toolchain, as it contains features that significantly improve the DX or are critical for some purposes. I also use `nightly` for my personal projects.

**Comment by kchibisov** [comment]
@ShikiSuen please, mind your language, even though it seems like you've deleted/edited a bunch of replies, it doesn't mean that I haven't seen them. This is not the first time when you behave _impulsive_, which may be off putting for certain people.

>Some Rust users prefer the nightly toolchain, as it contains features that significantly improve the DX or are critical for some purposes. I also use nightly for my personal projects.

Stable rustfmt can not do good formatting unfortunately and it's like that for a very long time. As you can see, we use nightly only for fmt, and not for anything else.

If you have major issues with formatting, you should politely ask that you can not do that, instead of exaggerating.

**Comment by kchibisov** [comment]
>The nightly rustfmt hinders the TODO to be in a standalone line. Is there a way to stop the nightly rustfmt from doing this stupid?

The general rule is to separate paragraphs with a blank line. I also don't like `TODO` in code in general, but it's up to @madsmtm to complain about such things, since it's not my backend.

**Comment by ShikiSuen** [comment]
@kchibisov 

If I were you, I'll clearly write these things in `README.md` prior to ask the contributor to use cargo-fmt:

1. Requirements about nightly toolchain and nightly cargo-fmt.
2. Detailed instructions regarding how to install and run them to help stable-only Rust users to stay on the right track.

What I experienced today is of extreme frustration.

I got blamed by those users of my input method app 
... because of all IMKTextInput client apps
... written in Rust directly or indirectly using this package.
... And I come here trying to fix the problem here to stop being blamed by my users.

And what I have experienced here today is totally unfair to me.

I'm sorry for what you have said "impulsive", but I really need to take a break.

(You might have felt something else impulsive from my previous comments. That's all because I got blamed by something that is technically not my fault... Especially the buggy InputMethodKit in macOS.)

**Comment by ShikiSuen** [comment]
@kchibisov 
I added the formatting policy to the README.md through this PR:
https://github.com/rust-windowing/winit/pull/3680

**Comment by kchibisov** [comment]
>Detailed instructions regarding how to install and run them to help stable-only Rust users to stay on the right track.

The rust itself does educate on how to install the toolchains and we don't require a particular nightly, etc, it's just any nightly. Though, there's no section in `CONTRIBUTING.md` that nighly needed, indeed.

>And what I have experienced here today is totally unfair to me.

The thing is that this PR is not delayed because of new formatting change or anything like that, it's delayed because I want @madsmtm to look into it, and he is busy recently, even though I pinged him here numerous times. If PR is ready I'd just handle conflicts myself and merged on my own which is what I always do anyway, when I don't like something. It's just given that you decided to rebase recently, I decided to acknowledge that you in general should make your PR green (if they issue is on you on indicate that you don't know what to do).

>I got blamed by those users of my input method app... because of all IMKTextInput client apps written in Rust directly or indirectly using this package.... And I come here trying to fix the problem here to stop being blamed by my users.

If users are _blaming_ you for such thing, I'd just ignore them, because you don't **owe** anything them, unless you provide a _paid_ service for them. Then it's a bit a different matter. Like I understand that it could be hard if you're not experienced dealing with that, but you can not please everyone.

In this particular case you send a patch and then it's really _not your problem_ anymore. 

>(You might have felt something else impulsive from my previous comments. That's all because I got blamed by something that is technically not my fault... Especially the buggy InputMethodKit in macOS.)

It's a matter of the words you're using and how you react to things during the review, and it's not a language barrier or anything, because you need some knowledge to speak like that. Unless you've learned English by only talking in gaming discords or something, idk.

But besides all of that, keep in mind that winit in the end of the day is an open source project and we do that in our free time. I know that @madsmtm is not MIA, if they were, I'd handle that PR myself. Which requires more time from me than I want, etc, because I just don't like macOS in general.

**Comment by ShikiSuen** [comment]
(Re-review requests sent, conforming to the contribution guidelines.)

**Comment by kchibisov** [comment]
>(Re-review requests sent, conforming to the contribution guidelines.)

I mean, there's no need to ask for re-review given that review was not submitted in the first place...

**Review by kchibisov — APPROVED** [review]

**Review by pan93412 — APPROVED** [review]

**Review by madsmtm — APPROVED** [review]
I pushed a commit to simplify the implementation, the extra function wasn't really required especially when I changed the way we got the string (there was some extra round-tripping between `&str` and `NSString` that wasn't necessary).

Thanks @ShikiSuen, and apologies I didn't get to this sooner.

**Comment by ShikiSuen** [comment]
I agree with Madsmtm's amendment.
Is there anything unfinished prior to merging this PR?

**Comment by ShikiSuen** [comment]
> i.e. it's not unsound to pass a wrong index

Looks like it is just my lack of familiarity with Foundation working on programming languages other than Swift. In Swift it is always an asserted death if an out-of-bound index gets passed into an array.

**Comment by kchibisov** [comment]
>Swift it is always an asserted death if an out-of-bound index gets passed into an array.

yet, it's still fine to do so in swift, it'll just assert the same way. unsound would mean that safe code invokes undefined behavior, panic is not unsound.

**Comment by ShikiSuen** [comment]
> > Swift it is always an asserted death if an out-of-bound index gets passed into an array.
> 
> yet, it's still fine to do so in swift, it'll just assert the same way. unsound would mean that safe code invokes undefined behavior, panic is not unsound.

Yes. What I want to say is that in such situation Swift doesn't throw an Exception for devs to handle. It just crashes instead. The generated IPS crash report in macOS console indicates that it is caused by array-out-of-bounds.

**Comment by madsmtm** [comment]
Thanks, and apologies that it took so long to get merged, I kept forgetting about it!

**Pull request merged**
